### PR TITLE
Make TLCCache work in a constant context

### PR DESCRIPTION
**Pull request by fhackett**
It's not a fundamental fix for `TLCEval`, but this addresses #1212 for me. As described in the issue, I was unable to make TLC consistently cache a value. `TLCCache` only worked for variable-level formulas, and `TLCEval` silently fails depending on how the definition is referenced (e.g., it stops working if you refer to your constant from inside a set comprehension). Given that `TLCEval` is very hard to fix, I followed Markus's advice to pivot.

Instead of working on `TLCEval`, I extended `TLCCache` to work in constant contexts. This means, as long as you can make a sufficiently unique "closure" object that identifies your value, you can cache anything you want at the constant or variable levels (no applications for other levels come to mind yet).

The intended application is in combination with IOUtils, to ensure your datafile is not silently reloaded in a loop.

Inventory of changes:
- `TLCCache` has a 2nd code path, where values are cached in the provided `Tool`. I do this by adding a concurrent hash map, so that multiple workers can safely use this code path, and I don't have to worry about optimizing read and write concurrency management.
- A test that uses `TLCGet`/`TLCSet` to confirm a cached value is computed only once, from assumptions, to init, to next-state.

One thing I'm assuming: it's sufficient to call `.normalize()` on a value you're going to cache across workers, so that you can safely reference it across said workers without accidentally doing thread unsafe lazy evaluation. I think that's probably reasonable, from how I understand `.normalize()`, but do correct me if I'm wrong.

### CODE REVIEW

File: tlatools/org.lamport.tlatools/src/tlc2/module/TLCExt.java [hunk 1]
```diff
@@ -333,8 +333,12 @@ public static Value tlcEval2(final Tool tool, final ExprOrOpArgNode[] args, fina
 		final ExprOrOpArgNode expr = args[0];
 		final ExprOrOpArgNode closure = args[1];
 
-		if (expr.getLevel() == LevelConstants.VariableLevel) {
-
+		if (expr.getLevel() == LevelConstants.ConstantLevel) {
+			final Value key = tool.eval(closure, c, s0, s1, control, cm).normalize();
+			return tool.getOrSetCached(key, () -> {
```
Comment (lemmy) [inline tlatools/org.lamport.tlatools/src/tlc2/module/TLCExt.java]: Have you considered `tla2sany.semantic.SemanticNode.setToolObject(int, Object)` instead of introducing a Map in Tool?

Comment (fhackett) [reply tlatools/org.lamport.tlatools/src/tlc2/module/TLCExt.java]: No I hadn't - I tried to understand how that API works, and it seems it allows storing run specific data in some AST node, if you have a unique ID for it.

In principle I can imagine storing the cache as such a tool object (I can also see how to allocate a specific ID), but I'm confused - maybe I misunderstood how thread safety is done here. Are tool objects / ASTs isolated per thread / worker? If not, those tool objects are unprotected array manipulations, and given TLCCache would get called for the first time at any point, I couldn't guarantee to not do bad things when initializing the cache.

Related: I tried some more testing "live" in TraceLink, and found (1) a bug in this PR, that somehow the field can still be null, and (2) some perf questions, where the concurrent map might be slowing somethings down. Working on those regardless.

Comment (lemmy) [reply tlatools/org.lamport.tlatools/src/tlc2/module/TLCExt.java]: Yes, you would need to introduce explicit locking in `TLCCache`, similar to how it’s done in `TLCEval.java`.  The `ReaderWriterLock` should provide sufficient concurrency.

Comment (fhackett) [reply tlatools/org.lamport.tlatools/src/tlc2/module/TLCExt.java]: I was part way through writing a message explaining why I thought that was unsafe, but I changed my mind.
My original comment was about how if TLCEval and TLCCache both have their own lock that protects the array underlying `setToolObject` , then surely you have 2 locks protecting 1 thing, and that will lead to trouble.

I actually think it's fine, but the reasoning is subtle. Any of these private locks protect specifically the AST "under" the `TLCEval(...)` of `TLCCache(...)` call. If I have the nonsensical but possible combination `TLCCache(TLCEval(???), {})`, and both caching mechanisms activate at the same time, we are still protecting 2 arrays with 2 locks, relatively speaking, because we can never be sharing the same AST node.

In that case, I think I get it. I pushed an alternative version that fixes a bug I found and also follows your suggestion.

File: tlatools/org.lamport.tlatools/src/tla2sany/StandardModules/TLCExt.tla [hunk 2]
```diff
@@ -168,8 +168,7 @@ TLCCache(expression, closure) ==
   (* Equals the value that an earlier evaluation of the expression evaluated to *)
   (* iff the closure of the earlier evaluation and this evaluation are equal.   *)
   (*                                                                            *)
-  (* Currently only implemented for state-level formulas.  See TLC!TLCEval for  *)
-  (* constant-level formulas.                                                   *)
+  (* Works in constant and variable-level contexts.                             *)
```
Comment (lemmy) [inline tlatools/org.lamport.tlatools/src/tla2sany/StandardModules/TLCExt.tla]: "state-level" is more commonly used compared to "variable-level" .

File: tlatools/org.lamport.tlatools/src/tlc2/module/TLCExt.java [hunk 3]
```diff
@@ -326,15 +328,64 @@ public static synchronized Value tlcModelValue(final Value val) {
 		return ModelValue.add(str.val.toString());
 	}
 	
+	private static final StampedLock tlcEval2Lock = new StampedLock();
```
Comment (lemmy) [inline tlatools/org.lamport.tlatools/src/tlc2/module/TLCExt.java]: Could you explain why a stamped instead of a ReaderWriteLock?

Comment (fhackett) [reply tlatools/org.lamport.tlatools/src/tlc2/module/TLCExt.java]: I read about its read locking (by far the most common case here) being faster than ReaderWriterLock, something about recursive locking support / the cache effect of timestamps. But also, it looked interesting and I wanted to try it.

While I don't have a strong opinion, I am now a little curious if the difference is measurable at all between the two locking methods. I have some specs that make extensive use of `TLCCache`, so there is some merit in empirically checking if this matters. I'll reply with results when I get back to this.

Comment (lemmy) [reply tlatools/org.lamport.tlatools/src/tlc2/module/TLCExt.java]: The performance difference has to be noticeable to justify the additional complexity arounds `StampLock`'s API compared to `ReaderWriterLock`.

File: tlatools/org.lamport.tlatools/src/tlc2/module/TLCExt.java [hunk 4]
```diff
@@ -326,15 +328,64 @@ public static synchronized Value tlcModelValue(final Value val) {
...
+				}
+				if (value != null) {
+					return value;
+				}
+				
+				tlcEval2Lock.unlock(stamp);
+				stamp = 0;
+				
+				// Note: do this outside of the write lock, otherwise recursive TLCCache invocations that miss
+				// the cache will deadlock trying to write to the cache while an outer call holds the lock.
```
Comment (lemmy) [inline tlatools/org.lamport.tlatools/src/tlc2/module/TLCExt.java]: Reentrancy wouldn't be an issue with `ReentrantReadWriteLock`.

File: tlatools/org.lamport.tlatools/test-model/ConstantContextTLCCache.tla [hunk 5]
```diff
@@ -0,0 +1,29 @@
+---- MODULE ConstantContextTLCCache ----
+EXTENDS Integers, TLC, TLCExt
+
+COUNT_ID == 1
+
+ASSUME TLCSet(COUNT_ID, 0)
```
Comment (lemmy) [inline tlatools/org.lamport.tlatools/test-model/ConstantContextTLCCache.tla:10]: Consider adding a couple of lines to explain what's going on in this test.

File: tlatools/org.lamport.tlatools/test/tlc2/module/ConstantContextTLCCacheTest.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
+package tlc2.module;
```
Comment (lemmy) [inline tlatools/org.lamport.tlatools/test/tlc2/module/ConstantContextTLCCacheTest.java:26]: Please add the usual MIT license header.

File: tlatools/org.lamport.tlatools/src/tlc2/module/TLCExt.java (hunk 3 above)
Comment (fhackett) [reply tlatools/org.lamport.tlatools/src/tlc2/module/TLCExt.java]: I checked. In a 5 minute 40 seconds run that calls `TLCCache` constantly, no difference is measurable. I changed it to the simpler lock.

**Review by lemmy — APPROVED** [review]
LGTM

Will merge soon unless @Calvin-L and/or @ahelwer also want to review.

File: tlatools/org.lamport.tlatools/src/tlc2/module/TLCExt.java [hunk 7]
```diff
@@ -326,15 +328,70 @@ public static synchronized Value tlcModelValue(final Value val) {
...
+				}
+				
+				// The element is still missing. Populate it (may recursively read and/or write lock)..
+				value = tool.eval(expr, c, s0, s1, control, cm);
+				// This is needed to ensure the value is usable from any thread going forward.
+				value.initialize();
+				
+				cache.put(key, value);
+				return value;
+			} finally {
```
Comment (lemmy) [inline tlatools/org.lamport.tlatools/src/tlc2/module/TLCExt.java:386]: I would hope for TLC to crash if anything goes south in between try/finally.  At least, `TLCEval` makes that assumption, and, thus, doesn't use try/finally.

Comment (fhackett) [reply tlatools/org.lamport.tlatools/src/tlc2/module/TLCExt.java:386]: I see. My main purpose for this is actually just code convenience, not necessarily exception safety. I wanted to use something like a try-with-resources, but locks don't support those. So, to avoid having to remember the unlock cases at each return, I just crammed them all in a finally block.

I can find a different structure if you like, but it might be harder to read. Or maybe this was just informational, in which case sure, that makes sense.

Comment (fhackett) [reply tlatools/org.lamport.tlatools/src/tlc2/module/TLCExt.java:386]: Actually I'm more confused after checking the code - `TLCEval` does also use try/finally https://github.com/tlaplus/tlaplus/blob/master/tlatools/org.lamport.tlatools/src/tlc2/module/TLCEval.java#L114

I didn't think I entirely made up the pattern... anyhow, let me know if you recommend changing something, and/or what you meant to say. I might be missing context.
